devito/finite_differences/differentiable.py

```diff
@@ -469,6 +470,91 @@ class Mod(DifferentiableOp, sympy.Mod):
     __sympy_class__ = sympy.Mod
 
 
+class IndexSum(DifferentiableOp):
+
+    """
+    A tensor contraction over one or more Dimensions.
+    """
+
+    is_commutative = True
+
+    def __new__(cls, expr, dimensions, **kwargs):
+        dimensions = as_tuple(dimensions)
+        if not dimensions:
+            return expr
+        for d in dimensions:
+            try:
+                if d.is_Dimension and is_integer(d.symbolic_size):
+                    continue
+            except AttributeError:
+                pass
+            raise ValueError("Expected Dimension with numeric size, "
+                             "got `%s` instead" % d)
+        for d in dimensions:
+            if d not in expr.free_symbols:
+                raise ValueError("Dimension `%s` must appear in `expr`" % d)
+        for i in expr.find(IndexSum):
+            for d in dimensions:
+                if d in i.dimensions:
+                    raise ValueError("Dimension `%s` already appears in a "
+                                     "nested tensor contraction" % d)
+
+        obj = sympy.Expr.__new__(cls, expr, *dimensions)
+        obj._expr = expr
+        obj._dimensions = dimensions
+
+        return obj
+
+    def __repr__(self):
+        return "IndexSum(%s, (%s))" % (self.expr,
+                                       ', '.join(d.name for d in self.dimensions))
+
+    __str__ = __repr__
+
+    @property
+    def expr(self):
+        return self._expr
+
+    @property
+    def dimensions(self):
+        return self._dimensions
+
+    @property
+    def evaluate(self):
+        # Overrides sympy.Expr.expand
+        values = product(*[list(range(d._min, d._max + 1)) for d in self.dimensions])
+        terms = []
+        for i in values:
+            mapper = dict(zip(self.dimensions, i))
+            terms.append(self.expr.xreplace(mapper))
+        return sum(terms)
+
+
+class IndexDerivative(IndexSum):
```

Comment:
Index or Indexed? Not sure need "Derivative" no? This is just a standard "stencil expression"

Reply:
It needs "Derivative" because it will ultimately be used for unexpanded derivatives in place of the fully unrolled Adds (`EvalDerivative`) that we generate today.

I'm honestly not sure about the name, but since this is far from the final state, I think we can postpone or discuss offline on Slack to converge to a better name